Mathlib/Combinatorics/SimpleGraph/Copy.lean

```diff
@@ -0,0 +1,203 @@
+
+section Copy
+
+/-- The type of copies as a subtype of *injective* homomorphisms. -/
+abbrev Copy (A : SimpleGraph α) (B : SimpleGraph β) :=
+  { f : A →g B // Function.Injective f }
+
+/-- An injective homomorphism gives rise to a copy. -/
+abbrev Hom.toCopy (f : A →g B) (h : Function.Injective f) : Copy A B := ⟨f, h⟩
+
+/-- An embedding gives rise to a copy. -/
+abbrev Embedding.toCopy (f : A ↪g B) : Copy A B := Hom.toCopy f.toHom f.injective
+
+/-- An isomorphism gives rise to a copy. -/
+abbrev Iso.toCopy (f : A ≃g B) : Copy A B := Embedding.toCopy f.toEmbedding
+
+namespace Copy
+
+/-- A copy gives rise to a homomorphism. -/
+abbrev toHom : Copy A B → A →g B := Subtype.val
+
+@[simp] lemma coe_toHom (f : Copy A B) : ⇑f.toHom = f := rfl
+
+lemma injective : (f : Copy A B) → (Function.Injective f.toHom) := Subtype.prop
+
+instance : FunLike (Copy A B) α β where
+  coe f := DFunLike.coe f.toHom
+  coe_injective' _ _ h := Subtype.val_injective (DFunLike.coe_injective h)
+
+@[simp] lemma coe_toHom_apply (f : Copy A B) (a : α) : ⇑f.toHom a = f a := rfl
+
+/-- A copy induces an embedding of edge sets. -/
+def mapEdgeSet (f : Copy A B) : A.edgeSet ↪ B.edgeSet where
+  toFun := Hom.mapEdgeSet f.toHom
+  inj' := Hom.mapEdgeSet.injective f.toHom f.injective
+
+/-- A copy induces an embedding of neighbor sets. -/
+def mapNeighborSet (f : Copy A B) (a : α) :
+    A.neighborSet a ↪ B.neighborSet (f a) where
+  toFun v := ⟨f v, f.toHom.apply_mem_neighborSet v.prop⟩
+  inj' _ _ h := by
+    rw [Subtype.mk_eq_mk] at h ⊢
+    exact f.injective h
+
+/-- A copy gives rise to an embedding of vertex types. -/
+def asEmbedding (f : Copy A B) : α ↪ β := ⟨f, f.injective⟩
+
+/-- The identity copy from a simple graph to itself. -/
+@[refl] def refl (G : SimpleGraph V) : Copy G G := ⟨Hom.id, Function.injective_id⟩
+
+/-- The copy from a subgraph to the supergraph. -/
+def ofLE {G₁ G₂ : SimpleGraph V} (h : G₁ ≤ G₂) : Copy G₁ G₂ :=
+  ⟨Hom.ofLE h, Function.injective_id⟩
+
+/-- The copy from an induced subgraph to the initial simple graph. -/
+def induce (G : SimpleGraph V) (s : Set V) : Copy (G.induce s) G :=
+  (Embedding.induce s).toCopy
+
+/-- The composition of copies is a copy. -/
+def comp (g : Copy B C) (f : Copy A B) : Copy A C := by
+  use g.toHom.comp f.toHom
+  rw [Hom.coe_comp]
+  exact Function.Injective.comp g.injective f.injective
+
+@[simp]
+theorem comp_apply (g : Copy B C) (f : Copy A B) (a : α) : g.comp f a = g (f a) :=
+  RelHom.comp_apply g.toHom f.toHom a
+
+end Copy
+
+/-- A `Subgraph G` gives rise to a copy from the coercion to `G`. -/
+def Subgraph.coeCopy (G' : G.Subgraph) : Copy G'.coe G :=
+  G'.hom.toCopy Subgraph.hom.injective
+
+end Copy
+
+section IsIsoSubgraph
+
+/-- The relation `IsIsoSubgraph A B` says that `B` contains a copy of `A`.
+
+This is equivalent to the existence of an isomorphism from `A` to a subgraph of `B`. -/
+abbrev IsIsoSubgraph (A : SimpleGraph α) (B : SimpleGraph β) := Nonempty (Copy A B)
```

Comment:
Could you adopt [my `IsContained` API](https://github.com/YaelDillies/LeanCamCombi/blob/master/LeanCamCombi/ExtrProbCombi/Containment.lean) here?